pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+			let price = match base_pool.share_price() {
+				Some(price) if price != fp!(0) => price,
+				_ => return false,
+			};
+			let mut budget = base_pool.free_stake + releasing_stake;
+			for request in &base_pool.withdraw_queue {
+				let withdraw_nft = Self::get_nft_attr(&base_pool, request.nft_id)
+					.expect("get nftattr should always success; qed.");
+				let amount = bmul(withdraw_nft.shares, &price);
+				if amount > budget {
+					// Run out of budget, let's check if the request is still in the grace period
+					return now - request.start_time > grace_period;
+				} else {
+					// Otherwise we allocate some budget to virtually fulfill the request
+					budget -= amount;
+				}
+			}
+			false
+		}
+
+		/// Mint a new nft in the Pool's collection and store some shares in it
+		pub fn add_stake_to_new_nft(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			userid: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> BalanceOf<T> {
+			let shares = match base_pool.share_price() {
+				Some(price) if price != fp!(0) => bdiv(amount, &price),
+				_ => amount, // adding new stake (share price = 1)
+			};
+			Self::mint_nft(base_pool, userid.clone(), shares, amount)
+				.expect("mint should always success; qed.");
+			base_pool.total_shares += shares;
+			base_pool.total_stake += amount;
+			base_pool.free_stake += amount;
+			shares
+		}
+
+		/// Remove some stakes from nft when withdraw or process_withdraw_queue called.
+		pub fn remove_stake_from_nft(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			shares: BalanceOf<T>,
+			nft: &mut NftAttr<BalanceOf<T>>,
+		) -> Option<(BalanceOf<T>, BalanceOf<T>, BalanceOf<T>)> {
+			let price = base_pool.share_price()?;
+			let amount = bmul(shares, &price);
+
+			let amount = amount.min(base_pool.free_stake).min(nft.stakes);
+
+			let user_shares = nft.shares.checked_sub(&shares)?;
+			let (user_shares, shares_dust) = extract_dust(user_shares);
+			let user_locked = nft.stakes.checked_sub(&amount)?;
+			let (user_locked, user_dust) = extract_dust(user_locked);
+
+			let removed_shares = shares + shares_dust;
+			let total_shares = base_pool.total_shares.checked_sub(&removed_shares)?;
+
+			let (total_stake, _) = extract_dust(base_pool.total_stake - amount);
+			if total_stake > Zero::zero() {
+				base_pool.free_stake -= amount;
+				base_pool.total_stake -= amount;
+			} else {
+				base_pool.free_stake = Zero::zero();
+				base_pool.total_stake = Zero::zero();
+			}
+			base_pool.total_shares = total_shares;
+			nft.shares = user_shares;
+			nft.stakes = user_locked;
+
+			Some((amount, user_dust, removed_shares))
+		}
+
+		#[frame_support::transactional]
+		pub fn mint_nft(
+			base_pool: &BasePool<T::AccountId, BalanceOf<T>>,
+			contributer: T::AccountId,
+			shares: BalanceOf<T>,
+			stakes: BalanceOf<T>,
+		) -> Result<NftId, DispatchError> {
+			Self::mint_nft_inner(base_pool.cid, contributer, shares, stakes)
+		}
+
+		#[frame_support::transactional]
```

Comment:
Why `transactional` for `_inner` function?

Reply:
the transactional outside is removed